File: binary_tree/129_sum_root_to_leaf_numbers.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sumNumbers(root: TreeNode | null): number {

    function dfs(node, curr){
        curr *= 10;
        curr += node.val

        if(!node.left && !node.right){
            return curr;
        }

        return ((node.left && dfs(node.left, curr)) || 0) + ((node.right && dfs(node.right, curr) || 0));
    }

    return dfs(root, 0)

};

File: binary_tree/129_sum_root_to_leaf_numbers.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn sum_numbers(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {

        pub fn dfs(node: Option<Rc<RefCell<TreeNode>>>, mut curr: i32) -> i32 {

            if let Some(node) = node {

                let node_b = node.borrow();

                curr *= 10;
                curr += node_b.val;

                if node_b.left.is_none() && node_b.right.is_none() {
                    return curr;
                }

                let left_val = if let Some(left) = &node_b.left {
                    dfs(Some(left.clone()), curr)
                } else {
                    0
                };

                let right_val = if let Some(right) = &node_b.right {
                    dfs(Some(right.clone()), curr)
                } else {
                    0
                };

                left_val + right_val

            } else {
                0
            }
        }

        dfs(root, 0)

    }
}